introduction_to_applying_machine_learning/deepar_chicago_traffic_violations/deepar_chicago_traffic_violations.ipynb

```diff
@@ -0,0 +1,467 @@
+    "        called before being able to use `predict`.\n",
+    "        \n",
+    "        Parameters:\n",
+    "        freq -- string indicating the time frequency\n",
+    "        prediction_length -- integer, number of predicted time points\n",
+    "       \n",
+    "       Return value: none.\n",
+    "       \"\"\"\n",
+    "        self.freq = freq\n",
+    "        self.prediction_length = prediction_length\n",
+    "        \n",
+    "    def predict(self, ts, cat=None, encoding='utf-8', num_samples=100, quantiles=['0.1', '0.5', '0.9']):\n",
+    "        \"\"\"Requests the prediction of for the time series listed in `ts`, each with the \n",
+    "        (optional) corresponding category listed in `cat`.\n",
+    "        \n",
+    "        Parameters:\n",
+    "        ts -- list of `pandas.Series` objects, the time series to predict\n",
+    "        cat -- list of integers (default: None)\n",
+    "        encoding -- string, encoding to use for the request (default: 'utf-8')\n",
+    "        num_samples -- integer, number of samples to compute at prediction time (default: 100)\n",
+    "        quantiles -- list of strings specifying the quantiles to compute (default: ['0.1', '0.5', '0.9'])\n",
+    "        \n",
+    "        Return value: list of `pandas.DataFrame` objects, each containing the predictions\n",
+    "        \"\"\"\n",
+    "        prediction_times = [x.index[-1]+1 for x in ts]\n",
+    "        req = self.__encode_request(ts, cat, encoding, num_samples, quantiles)\n",
+    "        res = super(DeepARPredictor, self).predict(req)\n",
+    "        return self.__decode_response(res, prediction_times, encoding)\n",
+    "    \n",
+    "    def __encode_request(self, ts, cat, encoding, num_samples, quantiles):\n",
+    "        instances = [series_to_obj(ts[k], cat[k] if cat else None) for k in range(len(ts))]\n",
+    "        configuration = {'num_samples': num_samples, 'output_types': ['quantiles'], 'quantiles': quantiles}\n",
+    "        http_request_data = {'instances': instances, 'configuration': configuration}\n",
+    "        return json.dumps(http_request_data).encode(encoding)\n",
+    "    \n",
+    "    def __decode_response(self, response, prediction_times, encoding):\n",
+    "        response_data = json.loads(response.decode(encoding))\n",
+    "        list_of_df = []\n",
+    "        for k in range(len(prediction_times)):\n",
+    "            prediction_index = pd.DatetimeIndex(start=prediction_times[k], freq=self.freq, periods=self.prediction_length)\n",
+    "            list_of_df.append(pd.DataFrame(data=response_data['predictions'][k]['quantiles'], index=prediction_index))\n",
+    "        return list_of_df\n",
+    "\n",
+    "\n",
+    "predictor = DeepARPredictor(endpoint=best_tuning_job_name,\n",
+    "                            sagemaker_session=sagemaker_session,\n",
+    "                            content_type=\"application/json\")\n"
```

Comment:
why is the content type "application/json" if the encoding/decoding uses pandas?

Reply:
nvm reread the code. ignore this comment 😅